patches/tModLoader/Terraria/ModLoader/Config/ConfigManager.cs

```diff
@@ -356,6 +478,206 @@ public static bool EnumerableEquals(IEnumerable a, IEnumerable b)
 		}
 		return !hasNextA && !hasNextB;
 	}
+
+	internal static string FormatTextAttribute(string key, string localization, object[] args)
+	{
+		// TODO: support FindKeyInScope-type key shortening? 
+		if (args == null)
+			return localization;
+		for (int i = 0; i < args.Length; i++) {
+			if (args[i] is string s && s.StartsWith("$"))
+				args[i] = Language.GetTextValue(s.Substring(1));
+		}
+		return Language.GetText(key).Format(args);
+	}
+
+	// Used to determine which key to register, based only on field/property, not class, not necessarily which key to use in UI.
+	internal static string GetLabelKey(PropertyFieldWrapper memberInfo, ModConfig config, bool fallbackToClass, bool throwErrors)
+	{
+		var label = (LabelKeyAttribute)Attribute.GetCustomAttribute(memberInfo.MemberInfo, typeof(LabelKeyAttribute));
+		if (label != null) {
+			if (label.malformed && throwErrors)
+				throw new ValueNotTranslationKeyException($"{nameof(LabelKeyAttribute)} only accepts localization keys for the 'key' parameter.");
+			return label.key;
+		}
+		label = (LabelKeyAttribute)Attribute.GetCustomAttribute(memberInfo.Type, typeof(LabelKeyAttribute));
+		if (label != null) {
+			if (label.malformed && throwErrors)
+				throw new ValueNotTranslationKeyException($"{nameof(LabelKeyAttribute)} only accepts localization keys for the 'key' parameter.", errorOnType: true);
+			if (fallbackToClass) // FinishSetup will catch errors on Label annotations on classes
+				return label.key;
+		}
+
+		// Autokey. If member belongs to a Type from this mod...
+		Type typeMemberBelongsTo = memberInfo.MemberInfo.DeclaringType;
+		if (typeMemberBelongsTo.Assembly == config.GetType().Assembly) { 
+			string keyName = config.Name;
+			// Use Type name if not ModConfig instance.
+			if (typeMemberBelongsTo != config.GetType())
+				keyName = typeMemberBelongsTo.Name;
+
+			return config.Mod.GetLocalizationKey($"Configs.{keyName}.{memberInfo.Name}.Label");
+		}
+		else {
+			// IsClass? IsValueType./ Ignore bool, int, etc...
+			string modName = typeMemberBelongsTo.Assembly.GetName().Name;
+			if (modName == "tModLoader" || modName == "FNA")
+				return $"Config.{typeMemberBelongsTo.Name}.{memberInfo.Name}.Label";
+			// Type is from some other mod.
+			return $"Mods.{modName}.Configs.{typeMemberBelongsTo.Name}.{memberInfo.Name}.Label";
+		}
+	}
+
+	internal static string GetLocalizedLabel(LabelKeyAttribute labelAttribute, PropertyFieldWrapper memberInfo) {
+		// Priority: Provided/Auto Key on member -> Key on class if member translation is empty string -> member name
+		var config = Interface.modConfig.pendingConfig;
```

Comment:
Not that I'm completely familiar with config loading, but using UI user state during loading code looks super confusing. I'm assuming you set some state in a `try finally` during load but...